src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -0,0 +1,443 @@
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, size_t count, char*& characters)
+{
+	if (!zip_ensure_valid_params (buf, buf_len, index, count))
+		return false;
+
+	characters = new char[count + 1];
+	memcpy (characters, buf + index, count);
+	characters [count] = '\0';
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_entry_info (uint8_t* buf, size_t buf_len, size_t& buf_offset, uint16_t& compression_method, uint32_t& local_header_offset, uint32_t& file_size, char*& file_name)
+{
+	static constexpr size_t CD_COMPRESSION_METHOD_OFFSET = 10;
+	static constexpr size_t CD_UNCOMPRESSED_SIZE_OFFSET  = 24;
+	static constexpr size_t CD_FILENAME_LENGTH_OFFSET    = 28;
+	static constexpr size_t CD_EXTRA_LENGTH_OFFSET       = 30;
+	static constexpr size_t CD_LOCAL_HEADER_POS_OFFSET   = 42;
+	static constexpr size_t CD_COMMENT_LENGTH_OFFSET     = 32;
+
+	size_t index = buf_offset;
+	if (!zip_ensure_valid_params (buf, buf_len, index, ZIP_CENTRAL_LEN))
+		return false;
+
+	uint8_t signature[4];
+	if (!zip_read_field (buf, buf_len, index, signature)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry signature");
+		return false;
+	}
+
+	if (memcmp (signature, ZIP_CENTRAL_MAGIC, sizeof(signature)) != 0) {
+		log_fatal (LOG_DEFAULT, "Invalid Central Directory entry signature");
+		return false;
+	}
+
+	index = buf_offset + CD_COMPRESSION_METHOD_OFFSET;
+	if (!zip_read_field (buf, buf_len, index, compression_method)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry 'compression method' field");
+		return false;
+	}
+
+	index = buf_offset + CD_UNCOMPRESSED_SIZE_OFFSET;;
+	if (!zip_read_field (buf, buf_len, index, file_size)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry 'uncompressed size' field");
+		return false;
+	}
+
+	uint16_t file_name_length;
+	index = buf_offset + CD_FILENAME_LENGTH_OFFSET;
+	if (!zip_read_field (buf, buf_len, index, file_name_length)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry 'file name length' field");
+		return false;
+	}
+
+	uint16_t extra_field_length;
+	index = buf_offset + CD_EXTRA_LENGTH_OFFSET;
+	if (!zip_read_field (buf, buf_len, index, extra_field_length)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry 'extra field length' field");
+		return false;
+	}
+
+	uint16_t comment_length;
+	index = buf_offset + CD_COMMENT_LENGTH_OFFSET;
+	if (!zip_read_field (buf, buf_len, index, comment_length)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry 'file comment length' field");
+		return false;
+	}
+
+	index = buf_offset + CD_LOCAL_HEADER_POS_OFFSET;
+	if (!zip_read_field (buf, buf_len, index, local_header_offset)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry 'relative offset of local header' field");
+		return false;
+	}
+	index += sizeof(local_header_offset);
+
+	if (file_name_length == 0) {
+		file_name = new char[1];
```

Comment:
Or should we just `assert()` that this doesn't happen?  What's it *mean* to encounter a filename with no length?

Reply:
We pass `file_name` to a number of other functions which may or may not accept a `nullptr`. Using an empty string for a name in a very unlikely situation like here makes for simpler and shorter code later on - instead of checking for `nullptr` we simply pass the name to other functions and let them do their thing.

Reply:
An empty file name is valid in ZIP - it means the data came from [standard input](https://github.com/nih-at/libzip/blob/master/docs/appnote.txt#L874)

Reply:
That's good to know.  TIL!